### DISCUSSION TITLE
Unknown column '' in 'field list', using MySQL

### DISCUSSION THREAD

**Discussion by Fogedge** [Budibase/budibase#4960] [Help]
Hi, I use Budibase with MySQL. 
I am completely new and Budibase is easy to use and great. So far I have managed to make one app which is really great. This is testimony to Budibase being very easy to use, even to inexperienced people!

But with my next app I ran into a problem. 

I have a table with many foreign keys (one to many keys) to several parent tables. In the second app I joined a child table to this table so it became also a parent table while still being a child table to several parent tables.
Now, when I want to update any of the tables using the :ID forms I get the error message: "Unknown column '' in 'field list'". The 'Unknown column' referred to is the id column of the table I want to update (using the save button). When pressing the save button the table actually updates correctly with the new value, despite the error message, but it is annoying to get this error message and the form does not return to the table view.  

If I join the tables with many to many join tables the tables update correctly without the error message. So this can solve the problem, apparently. But I don't like having to use many to many join tables when I could just join the parent and child tables with simple one to many relationships. 

It seems this problem started when I added the child table in the second app. Can this be what causes the problem - does Budibase / MySQL get 'confused', with the child table having many parents and then becoming a parent itself? 
Is it OK to solve the problem by using many to many relationships instead of one to many - or do I risk to run in to trouble later if I have not solved the fundamental reason for the problem?

The first app still updates fine without any error messages on the same tables but without being joined to the added child table in the second app.  
In the second app I also created a master detail form with the added child table, using the instructions in Budibase docs (which have since been removed). It works fine. 

I have checked that all the keys between the tables are matching properly, that the foreign keys are set to cascade and all are visible, and that the foreign keys are placed correctly in the child tables as they are supposed to be.

**Comment by mjashanks** [comment]
Hi @Fogedge would you be able to post your `CREATE TABLE` scripts for all tables in the model. I will try and reproduce your issue.

If you're not comfortable posting that in here, feel free to send to support@budibase.com

**Reply by Fogedge** [reply]
> Hi @Fogedge would you be able to post your `CREATE TABLE` scripts for all tables in the model. I will try and reproduce your issue.
> 
> If you're not comfortable posting that in here, feel free to send to [support@budibase.com](mailto:support@budibase.com)

Hi. 
I have created a small simple database. 

![image](https://user-images.githubusercontent.com/96333242/170036919-425adf8e-0b83-45f7-bd18-be0b9d8c2b7f.png)

I still get the problem. 
The referencing table can update/save without problems (after I pressed the save button in the /ghi/:id form and then returned to the table view). 

![image](https://user-images.githubusercontent.com/96333242/170037151-1599b3c5-2067-4546-946d-179483370d9a.png)

The referenced tables give the error message when trying to save a record. 

![image](https://user-images.githubusercontent.com/96333242/170037410-e9f11b58-f20c-44c8-a411-dd4e86c82197.png)

This happens for all the 3 referenced tables (but if the box for the 'ghi' referenced field is empty it appears that it will save properly) . 

Below are the 'CREATE TABLE' scripts for all 4 tables. 
Can you try to reproduce with these? 

CREATE TABLE `abc` (
  `idABC` int NOT NULL AUTO_INCREMENT,
  `ABCcreate` datetime DEFAULT CURRENT_TIMESTAMP,
  `ABCmodify` datetime DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  `ABC1` varchar(45) DEFAULT NULL,
  `ABC2` varchar(45) DEFAULT NULL,
  PRIMARY KEY (`idABC`)
) ENGINE=InnoDB AUTO_INCREMENT=7 DEFAULT CHARSET=utf8

CREATE TABLE `def` (
  `idDEF` int NOT NULL AUTO_INCREMENT,
  `DEFcreate` datetime DEFAULT CURRENT_TIMESTAMP,
  `DEFmodify` datetime DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  `DEF1` varchar(45) DEFAULT NULL,
  `DEF2` varchar(45) DEFAULT NULL,
  PRIMARY KEY (`idDEF`)
) ENGINE=InnoDB AUTO_INCREMENT=3 DEFAULT CHARSET=utf8

CREATE TABLE `jkl` (
  `idJKL` int NOT NULL AUTO_INCREMENT,
  `JKLcreate` datetime DEFAULT CURRENT_TIMESTAMP,
  `JKLmodify` datetime DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  `JKL1` varchar(45) DEFAULT NULL,
  `JKL2` varchar(45) DEFAULT NULL,
  PRIMARY KEY (`idJKL`)
) ENGINE=InnoDB AUTO_INCREMENT=4 DEFAULT CHARSET=utf8

CREATE TABLE `ghi` (
  `idGHI` int NOT NULL AUTO_INCREMENT,
  `GHIcreate` datetime DEFAULT CURRENT_TIMESTAMP,
  `GHImodify` datetime DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  `GHI1` varchar(45) DEFAULT NULL,
  `GHI2` varchar(45) DEFAULT NULL,
  `ABC_idABC` int NOT NULL DEFAULT '1',
  `DEF_idDEF` int NOT NULL DEFAULT '1',
  `JKL_idJKL` int NOT NULL,
  PRIMARY KEY (`idGHI`),
  KEY `fk_GHI_ABC1_idx` (`ABC_idABC`),
  KEY `fk_GHI_DEF1_idx` (`DEF_idDEF`),
  KEY `fk_GHI_JKL1_idx` (`JKL_idJKL`),
  CONSTRAINT `fk_GHI_ABC1` FOREIGN KEY (`ABC_idABC`) REFERENCES `abc` (`idABC`),
  CONSTRAINT `fk_GHI_DEF1` FOREIGN KEY (`DEF_idDEF`) REFERENCES `def` (`idDEF`),
  CONSTRAINT `fk_GHI_JKL1` FOREIGN KEY (`JKL_idJKL`) REFERENCES `jkl` (`idJKL`)
) ENGINE=InnoDB AUTO_INCREMENT=3 DEFAULT CHARSET=utf8

These are the relationships I have created in Budibase: 

![image](https://user-images.githubusercontent.com/96333242/170038889-4da86473-f915-4e19-8486-8950e0339679.png)

**Comment by Fogedge** [comment]
Hi Michael.
Thank you very much for your reply.
Please find attached the CREATE TABLE scripts.
Maybe I need to do a little bit more of explanation:


  *   I have made a complete sales system with excel sheets. It is working well but is heavy because of big sheets and repetition of data.
  *   Therefor I want to make a database system.
  *   I have already made the complete database table structure as I have envisioned it.
  *   And now I am trying to make the applications.
  *   So far I have only worked with basically 2 tables I have called shipments and inv_det and their related smaller tables
  *   The table shipments basically gives information about each container shipped. It is used for keeping a list of containers shipped from the supplier up-to-date. One container is one record in shipments.
  *   The table inv_det has the same information but in addition it details the items (the packages of goods) in a container.
  *   The remaining of the database will track the delivery of goods to shops as well as the sales to the customers and will calculate all the key figures such as turnover, costs and expenses, gross and net profit, etc.

I have been using MySQL Workbench to create the tables.

**Comment by kellis5137** [comment]
Did you ever figure out the problem? I recently upgraded and when I publish the app and try to say something with a particular mysql table, I get this:

```
**ERROR [1649597088769] (28 on ae1d7f4c15ef): Unknown column 'data' in 'field list'**
    Error: Unknown column 'data' in 'field list'
        at PromiseConnection.query (/app/node_modules/mysql2/promise.js:93:22)
        at MySQLIntegration.<anonymous> (/app/dist/integrations/mysql.js:125:56)
        at Generator.next (<anonymous>)
        at /app/dist/integrations/mysql.js:8:71
        at new Promise (<anonymous>)
        at __awaiter (/app/dist/integrations/mysql.js:4:12)
        at MySQLIntegration.internalQuery (/app/dist/integrations/mysql.js:119:20)
        at queryFn (/app/dist/integrations/mysql.js:213:53)
        at MySQLIntegration.<anonymous> (/app/dist/integrations/base/sql.js:456:36)
        at Generator.next (<anonymous>)
        at /app/dist/integrations/base/sql.js:8:71
        at new Promise (<anonymous>)
        at __awaiter (/app/dist/integrations/base/sql.js:4:12)
        at MySQLIntegration.queryWithReturning (/app/dist/integrations/base/sql.js:440:16)
        at MySQLIntegration.<anonymous> (/app/dist/integrations/mysql.js:214:39)
        at Generator.next (<anonymous>)
Trace: Error: Unknown column 'data' in 'field list'
    at PromiseConnection.query (/app/node_modules/mysql2/promise.js:93:22)
    at MySQLIntegration.<anonymous> (/app/dist/integrations/mysql.js:125:56)
    at Generator.next (<anonymous>)
    at /app/dist/integrations/mysql.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (/app/dist/integrations/mysql.js:4:12)
    at MySQLIntegration.internalQuery (/app/dist/integrations/mysql.js:119:20)
    at queryFn (/app/dist/integrations/mysql.js:213:53)
    at MySQLIntegration.<anonymous> (/app/dist/integrations/base/sql.js:456:36)
    at Generator.next (<anonymous>)
    at /app/dist/integrations/base/sql.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (/app/dist/integrations/base/sql.js:4:12)
    at MySQLIntegration.queryWithReturning (/app/dist/integrations/base/sql.js:440:16)
    at MySQLIntegration.<anonymous> (/app/dist/integrations/mysql.js:214:39)
    at Generator.next (<anonymous>) {
  code: 'ER_BAD_FIELD_ERROR',
  errno: 1054,
  sql: "**removed sql for security reasons**",
  sqlState: '42S22',
  sqlMessage: "Unknown column 'data' in 'field list'",
  expose: true,
  statusCode: 400,
  status: 400
}
```

I removed the sql, but I can tell you there is NO field name "data" that is within it. Funnier still, if I add a "data" column, it then goes "Unknown column 'type' in 'field list'"... Once again, there is no field named "type" that gets sent via the update sql statement... I can see the update in the logs and it had neither "data" or "type" fields.... is there something with the underlying ORM?
